### DISCUSSION TITLE
Custom factors and degrees of freedom.

### DISCUSSION THREAD

**Discussion by biggibig** [borglab/gtsam#1104] [Q&A] [answered]
Hello. I was wondering if there were any restriction on creating custom factors.

Let's say we are dealing with a 6DOF pose graph SLAM Rx Ry Rz, Tx Ty Tz. Can I make custom factors that are less than 6DOF? I know there is a GPS factor Tx Ty Tz that's only 3DOF. But For example can I make custom a factor consisting of 4DOF something like Rx Ry Rz Tz? Or some other reduced combination of the 6 degrees of freedom, maybe even a single DOF, for example if I know that my heading is really good?

**Comment by mikesheffler** [comment, marked as answer]
Yes, you can! GPS is a good example of affecting only a subset of the full 6DOF pose. Another is the [Relative Elevation factor](https://github.com/borglab/gtsam/blob/develop/gtsam_unstable/slam/RelativeElevationFactor.cpp), where `z`, if you will, is being affected, but rotations and _other_ changes in translation (`x`, and `y`) are not. Take a look at [the evaluateError method for that class](https://github.com/borglab/gtsam/blob/develop/gtsam_unstable/slam/RelativeElevationFactor.cpp#L20).

The only bit of warning is that _it is very easy to be wrong_ when calculating the Jacobians for your custom factor. Take a look at [an overview of some relevant math](https://github.com/borglab/gtsam/blob/develop/gtsam_unstable/slam/RelativeElevationFactor.cpp#L20) for hints on how to do your calculations. Check your work by putting together a unit test that uses GTSAM's numerical derivative capabilities; for example, here's the [unit test for the Relative Elevation factor](https://github.com/borglab/gtsam/blob/develop/gtsam_unstable/slam/tests/testRelativeElevationFactor.cpp).

Finally, consider expression factors. If you are clever about chaining together existing functions, you can sometimes write expression factors that are quite complicated without having to calculate Jacobians yourself at all. Look in [examples](https://github.com/borglab/gtsam/tree/develop/examples) for inspiration.

Good luck! Consider the [Google Group](https://groups.google.com/g/gtsam-users) as a resource well.

**Reply by biggibig** [reply]
That's great, lots of good resources to look at. Is it possible to make custom relative factors or only absolute? Relative elevation factor sounds like a relative factor just from it's name.  Thanks for the reply!

**Comment by mikesheffler** [comment]
> Is it possible to make custom relative factors or only absolute?

Did you mean "_is it possible to make **absolute** factors, or only **relative**_"? I wasn't sure from the sentence that came after it.

Anyway, both are possible!

_Relative_ stuff can be handy because there might be more than one thing that you are not sure about. So, `RelativeElevationFactor`, for example, is a _binary_ factor, because it deals with two things -- a `Pose3` and a `Point3`. You want to reason about the elevation of that `Pose3` with respect to that `Point3`, but the `Point3` is also a variable (you measured it, that measurement had some error, etc.).

_Absolute_ stuff is handy because ... maybe it's not relative. If you are thinking about just one variable, you're probably going to be using a _unary_ factor (one thing). The simplest example of a unary factor is a _prior_ factor. We use the term _prior_ when we're thinking about the _a priori_ distribution of a variable, but the mechanics of using the _prior_ factors available to us aren't fundamentally different than the mechanics of _unary_ factors -- they're the same thing, we just don't have two names for it. That's a liiiittle bit of a conceptual overload on that one word, but it's basically true.

Consider [this example of setting a prior from the Pose2 SLAM Example](https://github.com/borglab/gtsam/blob/develop/examples/Pose2SLAMExample.cpp#L73).

If what you mean by _absolute_ isn't the _unary_ part, but rather something closer to _this-quantity-isn't-up-for-debate_ (it's not a 'variable' at all), you could consider [Nonlinear Equality](https://github.com/borglab/gtsam/blob/develop/gtsam/nonlinear/NonlinearEquality.h), but realistically, you're usually inviting numerical problems if you go down that route, and you're better off using a traditional unary factor that has a very small uncertainty in the 'measurement'.

Be sure to check out [Frank's crash course guide to GTSAM](https://github.com/borglab/gtsam/blob/develop/doc/gtsam.pdf). 🙂

**Comment by biggibig** [comment]
Yes by absolute I meant priors. For some reason I assumed that gtsam only allowed custom unary factors. But if it allows custom relative as well that's even better. I'm going to have to review all the materials and update my gtsam version.
